File: src/main/scala/math/cat/SetMorphism.java
package math.cat;

import java.util.*;

/**
 * SetMorphism class: morphisms for sets.
 * All code is <a href="http://myjavatools.com/projects/Category/source.zip">here</a>
 */
public abstract class SetMorphism
    <X, // domain data type
     PX extends Set<X>, // domain type
     Y, // codomain data type
     PY extends Set<Y> // codomain type
    > extends Morphism<PX, PY> {

  /**
   * Constructor.
   *
   * @param domain domain of this morphism
   * @param codomain codomain of this morphism
   */
  public SetMorphism(PX domain, PY codomain) {
    super(domain, codomain);
    validate();
  }

  /**
   * Constructor.
   *
   * @param name name of this morphism
   * @param domain domain of this morphism
   * @param codomain codomain of this morphism
   */
  public SetMorphism(String name, PX domain, PY codomain) {
    super(name, domain, codomain);
    validate();
  }

  /**
   * Validates set morphism.
   * All we need is that, each domain element is mapped to a codomain element.
   */
  private void validate() {
    for (X x : domain()) assert(codomain().contains(apply(x))) : "Morphism value for " + x + " out of d1";
  }

  public String toString() {
    if (name() != null) return name();
    StringBuffer sb = new StringBuffer();
    for (X x : domain()) {
      if (sb.length() > 0) sb.append(", ");
      sb.append(x).append(" -> ").append(apply(x));
    }

    return "{" + sb + "}";
  }

  /**
   * Applies the set morphism, that is a function, to an element of the set
   * @param x element to apply the morphism to
   * @return morphism value on x
   */
  public abstract Y apply(X x);

  public int hashCode() {
    return name() != null ? name().hashCode() : (domain().hashCode() * 4/*random number*/ + codomain().hashCode());
  }

  /**
   * Two set morphisms are equal if they have equal domains and codomains and map domain elements to the same values.
   * Note that there's no negation in this calculation; there is a deep reason for it, to be disclosed much, much later.
   * 
   * @param other set morphism to compare
   * @return true if they are equal
   */
  public boolean equals(SetMorphism<X, PX, Y, PY> other) {
    boolean isEqual = domain().equals(other.domain()) &&
                      codomain().equals(other.codomain());
    for (X x : domain()) {
      isEqual = isEqual && apply(x).equals(other.apply(x));
    }
    return isEqual;
  }

  public boolean equals(Object o) { return o instanceof SetMorphism && (((SetMorphism)o).equals(this)); }

  /**
   * Factory method. Builds a set morphism, given two sets and a map.
   *
   * @param domain domain set
   * @param codomain codomain set
   * @param map maps domain elements to codomain elements
   * @return a new instance of SetMorphism
   */
  public static <X, Y>
  SetMorphism<X, Set<X>, Y, Set<Y>> Morphism(final Set<X> domain, final Set<Y> codomain, final Map<X, Y> map) {
    for (X x : domain) {
      assert map.containsKey(x) : "Map should be defined on element " + x;
    }

    return new SetMorphism<X, Set<X>, Y, Set<Y>>(domain, codomain) {
      public Y apply(X x) {
        assert domain.contains(x) : "Argument not in d0";
        return map.get(x);
      }
    };
  }

  /**
   * Factory method. Builds unit morphism for a set (identity function).
   *
   * @param s the set
   * @return identity morphism on the given set
   */
  public static <X> SetMorphism<X, Set<X>, X, Set<X>> unit(Set<X> s) {
    return new SetMorphism<X, Set<X>, X, Set<X>>(s, s) {
      public X apply(X x) {
        return x;
      }
    };
  }

  /**
   * Composes two set morphisms. Note that while SetMorphism can be subclassed to be used by morphisms of subclasses
   * of Set, composition cannot be defined in a generic way.
   * To make it possible in Java (with generics), one needs closures!
   *
   * @param f first morphism: X -> Y
   * @param g second morphism: Y -> Z
   * @return their composition g o f: X -> Z
   */
  public static <
      X, // domain data type for the first morphism
      Y, // codomain data type for the first morphism (same as domain type for the second morphism)
      Z  // codomain data type for the first morphism
     >
  SetMorphism<X, Set<X>, Z, Set<Z>> compose(
      final SetMorphism<X, Set<X>, Y, Set<Y>> f, // X -> Y
      final SetMorphism<Y, Set<Y>, Z, Set<Z>> g  // Y -> Z
    ) {
    assert f.codomain().equals(g.domain()): "Composition not defined";
    return new SetMorphism<X, Set<X>, Z, Set<Z>>(f.domain(), g.codomain()) {
      public Z apply(X x) {
        return g.apply(f.apply(x));
      }
    };
  }
}

File: src/main/scala/math/cat/SetMorphism.scala
package math.cat

import math.cat.SetMorphism._
import math.sets.Sets
import math.sets.Sets._
import scalakittens.Result
import scalakittens.Result._

/**
 * Morphism class for sets, and the accompanying object.
 */
class SetMorphism[X, Y] (
    val tag: String,
    val d0: Set[X],
    val d1: Set[Y],
    val function: X ⇒ Y)
  extends Morphism[Set[X], Set[Y]] with Map[X, Y] {

  override def toString: String = tag match {
    case "" ⇒ "{" + (d0 map (x ⇒ s"$x → ${this(x)}") mkString ", ")  + "}"
    case _  ⇒ s"$tag: ${d0.size} elements → ${d1.size} elements"
  }

  override def hashCode: Int = d0.hashCode * 4/*random number*/ + d1.hashCode
  
  /**
   * Two set morphisms are equal if they have equal d0s and d1s and map d0 elements to the same values.
   * Note that there's no negation in this calculation; there is a deep reason for it, to be disclosed much, much later.
   *
   * @param other set morphism to compare
   * @return true iff they are equal
   */
  def equals(other: SetMorphism[X, Y]): Boolean = {
    ((d0.equals(other.d0) && d1.equals(other.d1)) /: d0) ((eq, x) ⇒ eq & this(x) == other(x))
  }
  
  def compose[Z](g: SetMorphism[Y, Z]): Option[SetMorphism[X, Z]] = {
    OKif (d1 == g.d0) andThen build[X, Z](d0, g.d1, (x: X) ⇒ g(this(x))) asOption
  }
  
  def revert: SetMorphism[Y, Set[X]] = {
    val d1AsSet: Set[Y] = d1
    val d0AsSet: Set[X] = d0
    val d0ForRevert: Set[Set[X]] = Sets.powerset(d0AsSet)
    build[Y, Set[X]](
      d1AsSet,
      d0ForRevert,
      Sets.groupBy(d0AsSet, d1AsSet, this)) iHope
  }  
  
  override def contains(x: X): Boolean = d0.contains(x)

  def get(x: X): Option[Y] = if (contains(x)) Some(function(x)) else None

  override def size: Int = d0.size

  def - (x: X): Map[X, Y] = itsImmutable
  def +[Y1 >: Y] (kv: (X, Y1)): Map[X, Y] = itsImmutable
  override def updated[Y1 >: Y](x: X, y: Y1): Map[X, Y1] = itsImmutable
  def empty[PY] = throw new RuntimeException("No such thing exists as empty set morphism")
  def iterator: Iterator[(X, Y)] = d0.iterator map (x ⇒ (x, function(x)))
  def product: Set[Map[Y, X]] = exponent(d1, d0) filter(m ⇒ d1 forall {y ⇒ function(m(y)) == y})
  
  def restrictTo(newDomain: Set[X]): Result[SetMorphism[X, Y]] = {
    OKif(newDomain subsetOf d0) returning new SetMorphism[X, Y](tag, newDomain, d1, function)
  }
}

object SetMorphism {
  // Validates set morphism.
  // All we need is that each d0 element is mapped to a d1 element.
  def check[X, Y, T <: SetMorphism[X, Y]](f: T): Result[T] = Result.traverse {
    for {x <- f.d0} yield {
      val y = f.function(x)
      val yInD1 = f.d1 contains y
      OKif(yInD1, s"${f.tag}: Value $y for $x should be in d1 ${f.d1}")
    }
  } returning f
  
  def build[X, Y](d0: Set[X], d1: Set[Y], function: X ⇒ Y): Result[SetMorphism[X, Y]] =
    build[X, Y]("", d0, d1, function)

  def build[X, Y](name: String, d0: Set[X], d1: Set[Y], function: X ⇒ Y): Result[SetMorphism[X, Y]] =
    check[X, Y, SetMorphism[X, Y]](new SetMorphism[X, Y](name, d0, d1, function))

  def inclusion[X](d0: Set[X], d1: Set[X]): Result[SetMorphism[X, X]] = {
    OKif(d0 subsetOf d1) returning new SetMorphism("⊂", d0, d1, identity)
  }
  
  def id[X](domain: Set[X]) = new SetMorphism[X, X]("1", domain, domain, x ⇒ x)

  def const[X, Y](d0: Set[X], d1: Set[Y], value: Y) =
    new SetMorphism[X, Y]("`" + value + "`", d0, d1, _ ⇒ value)

  def hom[X, Y](xs: Set[X], ys: Set[Y]): Set[SetMorphism[X, Y]] = {
      val maps = Sets.exponent(xs, ys)
      val morphisms = maps flatMap (build(xs, ys, _).asOption)
      def predicate(m: SetMorphism[X, Y]) = m.d0 == xs && m.d1 == ys && maps.contains(m)
      setOf(morphisms, maps.size, predicate _)
  }

  def pullback[X, Y, Z](f: SetMorphism[X, Z], g: SetMorphism[Y, Z]): (SetMorphism[(X, Y), X], SetMorphism[(X, Y), Y]) = {
    val pullbackSet = Sets.pullback(f.d0, g.d0, f, g)
    val p0 = (p: (X, Y)) ⇒ p._1
    val p1 = (p: (X, Y)) ⇒ p._2
    (build(pullbackSet, f.d0, p0) andAlso build(pullbackSet, g.d0, p1)) iHope
  }
}